pkg/loghttp/query.go

```diff
@@ -269,19 +270,31 @@ func ParseRangeQuery(r *http.Request) (*RangeQuery, error) {
 		return nil, err
 	}
 
-	result.Step, err = step(r, result.Start, result.End)
+	result.Step, err = step(r)
 	if err != nil {
 		return nil, err
 	}
-
-	if result.Step <= 0 {
+	if result.Step < 0 {
 		return nil, errNegativeStep
 	}
 
-	// For safety, limit the number of returned points per timeseries.
-	// This is sufficient for 60s resolution for a week or 1h resolution for a year.
-	if (result.End.Sub(result.Start) / result.Step) > 11000 {
-		return nil, errStepTooSmall
+	// Additional rules are required for the step on metric queries,
+	// so we need to parse the expression now to see what type it is.
+	expr, err := logql.ParseExpr(result.Query)
+	if err != nil {
+		return nil, err
+	}
+	switch expr.(type) {
+	case logql.SampleExpr:
+		if result.Step == 0 {
+			result.Step = time.Duration(defaultQueryRangeStep(result.Start, result.End)) * time.Second
+		}
+
+		// For safety, limit the number of returned points per timeseries.
+		// This is sufficient for 60s resolution for a week or 1h resolution for a year.
+		if (result.End.Sub(result.Start) / result.Step) > 11000 {
+			return nil, errStepTooSmall
+		}
```

Comment:
I think this is fine as-is, but you might want to consider inverting this condition: always default to non-zero unless it's a log expr, since the log expr is the exceptional case here.

Reply:
I don't disagree but the code would be more convoluted to invert the case, you'd have to somehow mark the step value was 0, then apply the default code that makes it non-zero, then check if it's a loq query and the initial value before applying default was zero and force it to zero again, that feels weird to me.

Reply:
Yeah, I understand. Maybe just a comment noting that if a new case is added to keep the step in mind?

Reply:
hrm, what about adding in the log case as a no-op and a default case that panics?

Reply:
:+1: good idea